mailpoet/mailpoet
# Refactor MPAPI subscribeToLists to Doctrine [MAILPOET-3813]

This PR refactors 

- \MailPoet\API\MP\v1\API::subscribeToList 
- \MailPoet\API\MP\v1\API::subscribeToLists 

The methods are now in a new class `lib/API/MP/v1/Subscribers.php`
Tests have been moved and refactored from `tests/integration/API/MP/APITest.php` to `tests/integration/API/MP/SubscribersTest.php`

I have reused the class `lib/API/JSON/ResponseBuilders/SubscribersResponseBuilder.php` to build the response replacing 
`->withCustomFields()->withSubscriptions()->asArray();`
I have added a parameter to the function 'build' to add more data on subscriptions if we are calling it from MPAPI. 
This information is redundant but I wanted to avoid modifying current MPAPI response, [as documented here](https://github.com/mailpoet/mailpoet/blob/master/doc/api_methods/GetSubscriber.md).

There are two functions where the old models are still present that require refactoring other components.
* sendSubscriberNotification
* _sendConfirmationEmail : The old models on this function can be refactored with this PR: https://github.com/mailpoet/mailpoet/pull/3834

I have also replaced many of the fixtures for Subscribers and Segments on APITest.php with the Doctrine version.

To test:
```
./do --test "test:integration --file=tests/integration/API/MP/APITest.php"
./do --test "test:integration --file=tests/integration/API/MP/SubscribersTest.php"
```
You can also check that the response structure has not changed after the refactoring.
I haven't tested the integration with a 3rd party plugin but if you know of a good candidate, I wouldn't mind doing it.

[MAILPOET-3813]

[MAILPOET-3820]: https://mailpoet.atlassian.net/browse/MAILPOET-3820?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[MAILPOET-3813]: https://mailpoet.atlassian.net/browse/MAILPOET-3813?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
